api/src/exam-environment/routes/exam-environment.ts

```diff
@@ -0,0 +1,578 @@
+ */
+export const examEnvironmentOpenRoutes: FastifyPluginCallbackTypebox = (
+  fastify,
+  _options,
+  done
+) => {
+  fastify.post(
+    '/exam-environment/token/verify',
+    {
+      schema: schemas.examEnvironmentTokenVerify
+    },
+    tokenVerifyHandler
+  );
+  done();
+};
+
+interface JwtPayload {
+  examEnvironmentAuthorizationToken: string;
+}
+
+/**
+ * Verify an authorization token has been generated for a user.
+ *
+ * Does not require any authentication.
+ *
+ * **Note**: This has no guarantees of which user the token is for. Just that one exists in the database.
+ */
+async function tokenVerifyHandler(
+  this: FastifyInstance,
+  req: UpdateReqType<typeof schemas.examEnvironmentTokenVerify>,
+  reply: FastifyReply
+) {
+  const { 'exam-environment-authorization-token': encodedToken } = req.headers;
+
+  try {
+    jwt.verify(encodedToken, JWT_SECRET);
+  } catch (e) {
+    // TODO: What to send back here? Request is valid, but token is not?
+    void reply.code(200);
+    return reply.send(
+      ERRORS.FCC_EINVAL_EXAM_ENVIRONMENT_AUTHORIZATION_TOKEN(JSON.stringify(e))
+    );
+  }
+
+  const payload = jwt.decode(encodedToken) as JwtPayload;
+
+  const examEnvironmentAuthorizationToken =
+    payload.examEnvironmentAuthorizationToken;
+
+  const token = await this.prisma.examEnvironmentAuthorizationToken.findFirst({
+    where: {
+      id: examEnvironmentAuthorizationToken
+    }
+  });
+
+  if (!token) {
+    void reply.code(200);
+    return reply.send({
+      data: 'Token does not appear to have been created.'
+    });
+  } else {
+    void reply.code(200);
+    return reply.send({
+      data: {
+        createdDate: token.createdDate
+      }
+    });
+  }
+}
+
+/**
+ * Generates an exam for the user.
+ *
+ * Requires token to be validated and TODO: live longer than the exam attempt.
+ */
+async function postExamGeneratedExamHandler(
+  this: FastifyInstance,
+  req: UpdateReqType<typeof schemas.examEnvironmentPostExamGeneratedExam>,
+  reply: FastifyReply
+) {
+  // Get exam from DB
+  const examId = req.body.examId;
+  // TODO: This throws if `id` is not serializable into an ObjectId
+  //       Should this be caught, and thrown to the client as a client error?
+  const maybeExam = await mapErr(
+    this.prisma.envExam.findUnique({
+      where: {
+        id: examId
+      }
+    })
+  );
+  if (maybeExam.hasError) {
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT(JSON.stringify(maybeExam.error))
+    );
+  }
+
+  const exam = maybeExam.data;
+
+  if (!exam) {
+    void reply.code(404);
+    return reply.send(
+      ERRORS.FCC_ENOENT_EXAM_ENVIRONMENT_MISSING_EXAM('Invalid exam id given.')
+    );
+  }
+
+  // Check user has completed prerequisites
+  const { user } = req;
+  if (!user) {
+    void reply.code(500);
+    return reply.send(
+      ERRORS.FCC_ERR_EXAM_ENVIRONMENT('Unreachable. User not authenticated.')
+    );
+  }
```

Comment:
This is overly defensive, imo. The fact we're behind auth means the user exists. If, somehow, the user doesn't exist, then `checkPrerequisites` will throw when it attempts to, say, get `user.id`. At that point we investigate what on earth happened.

Reply:
I do not understand what your suggestion is? This check is here purely because TS does not know `user` _almost definitely should probably_ exist.